src/main/java/org/stellar/sdk/Sep10Challenge.java

```diff
@@ -40,4 +45,337 @@ public static String newChallenge(
+      throw new InvalidSep10ChallengeException("No verifiable signers provided, at least one G... address must be provided.");
+    }
+
+    // Verify all the transaction's signers (server and client) in one
+    // hit. We do this in one hit here even though the server signature was
+    // checked in the readChallengeTx to ensure that every signature and signer
+    // are consumed only once on the transaction.
+    Set<String> allSigners = new HashSet<String>(clientSigners);
+    allSigners.add(serverKeyPair.getAccountId());
+    Set<String> signersFound = verifyTransactionSignatures(transaction, allSigners);
+
+    // Confirm the server is in the list of signers found and remove it.
+    boolean serverSignerFound = signersFound.remove(serverKeyPair.getAccountId());
+
+    // Confirm we matched a signature to the server signer.
+    if (!serverSignerFound) {
+      throw new InvalidSep10ChallengeException(String.format("Transaction not signed by server: %s.", serverAccountId));
+    }
+
+    // Confirm we matched signatures to the client signers.
+    if (signersFound.isEmpty()) {
+      throw new InvalidSep10ChallengeException("Transaction not signed by any client signer.");
+    }
+
+    // Confirm all signatures were consumed by a signer.
+    if (signersFound.size() != transaction.getSignatures().size() - 1) {
+      throw new InvalidSep10ChallengeException("Transaction has unrecognized signatures.");
+    }
+
+    return signersFound;
+  }
+
+  /**
+   * Verifies that for a SEP-0010 challenge transaction
+   * all signatures on the transaction are accounted for and that the signatures
+   * meet a threshold on an account. A transaction is verified if it is signed by
+   * the server account, and all other signatures match a signer that has been
+   * provided as an argument, and those signatures meet a threshold on the
+   * account.
+   *
+   * @param challengeXdr    SEP-0010 transaction challenge transaction in base64.
+   * @param serverAccountId Account ID for server's account.
+   * @param network         The network to connect to for verifying and retrieving.
+   * @param threshold       The threshold on the client account.
+   * @param signers         The signers of client account.
+   * @return a list of signers that were found is returned, excluding the server account ID.
+   * @throws InvalidSep10ChallengeException If the SEP-0010 validation fails, the exception will be thrown.
+   * @throws IOException                    If read XDR string fails, the exception will be thrown.
+   */
+  public static Set<String> verifyChallengeTransactionThreshold(String challengeXdr, String serverAccountId, Network network, int threshold, Set<Signer> signers) throws InvalidSep10ChallengeException, IOException {
+    if (signers == null || signers.isEmpty()) {
+      throw new InvalidSep10ChallengeException("No signers provided.");
+    }
+
+    Map<String, Signer> accountIdSignerMap = new HashMap<String, Signer>();
```

Comment:
I think it would make the code slightly more readable to use a `Map<String, Integer>` to hold the weights for each signer:

```java
    Map<String, Integer> weightsForSigner = new HashMap<String, Integer>();
    for (Signer signer : signers) {
      weightsForSigner.put(signer.getKey(), signer.getWeight());
    }

    Set<String> signersFound = verifyChallengeTransactionSigners(challengeXdr, serverAccountId, network, weightsForSigner.keySet());

    int sum = 0;
    for (String signer : signersFound) {
      Integer weight = weightsForSigner.get(signer);
      if (weight != null) {
        sum += weight;
      }
    }

    if (sum < threshold) {
      throw new InvalidSep10ChallengeException(String.format("Signers with weight %d do not meet threshold %d.", sum, threshold));
    }
```